File: dataset/ZeroShot/CLAUDE_claude-sonnet-4-5-20250929/Task129.py

import os
import re
from pathlib import Path

class Task129:
    BASE_DIRECTORY = "./secure_files/"
    ALLOWED_EXTENSIONS = {".txt", ".json", ".xml", ".csv", ".log"}
    MAX_FILE_SIZE = 10 * 1024 * 1024  # 10MB
    SAFE_FILENAME_PATTERN = re.compile(r'^[a-zA-Z0-9_\\-\\.]+$')
    
    @staticmethod
    def retrieve_file(file_name):
        try:
            # Input validation
            if not file_name or not file_name.strip():
                return "Error: File name cannot be empty"
            
            file_name = file_name.strip()
            
            # Check for safe filename pattern
            if not Task129.SAFE_FILENAME_PATTERN.match(file_name):
                return "Error: Invalid file name format. Only alphanumeric, dash, underscore and dot allowed"
            
            # Check file extension
            extension = Path(file_name).suffix.lower()
            if extension not in Task129.ALLOWED_EXTENSIONS:
                return f"Error: File type not allowed. Allowed types: {Task129.ALLOWED_EXTENSIONS}"
            
            # Prevent path traversal attacks
            base_path = Path(Task129.BASE_DIRECTORY).resolve()
            file_path = (base_path / file_name).resolve()
            
            if not str(file_path).startswith(str(base_path)):
                return "Error: Path traversal detected. Access denied"
            
            # Check if file exists
            if not file_path.exists():
                return "Error: File not found"
            
            # Check if it's a file (not directory)\n            if not file_path.is_file():\n                return "Error: Invalid file type"\n            \n            # Check file size\n            if file_path.stat().st_size > Task129.MAX_FILE_SIZE:\n                return "Error: File size exceeds maximum allowed size"\n            \n            # Read file content\n            with open(file_path, 'r', encoding='utf-8') as f:\n                content = f.read()\n            \n            return f"Success: File retrieved\\nContent:\\n{content}"\n            \n        except IOError as e:\n            return f"Error: Unable to read file - {str(e)}"\n        except Exception as e:\n            return f"Error: Unexpected error - {str(e)}"\n    \n    @staticmethod\n    def setup_test_environment():\n        try:\n            os.makedirs(Task129.BASE_DIRECTORY, exist_ok=True)\n            \n            # Create test files\n            test_files = {\n                "test1.txt": "This is a test file 1",\n                "test2.json": '{"name": "test", "value": 123}',\n                "test3.csv": "id,name,age\\n1,John,30\\n2,Jane,25"\n            }\n            \n            for filename, content in test_files.items():\n                with open(os.path.join(Task129.BASE_DIRECTORY, filename), 'w') as f:
                    f.write(content)
        except Exception as e:
            print(f"Setup error: {str(e)}")

def main():
    Task129.setup_test_environment()
    
    print("=== Secure File Retrieval System ===\\n")
    
    # Test case 1: Valid file retrieval
    print("Test 1: Valid file retrieval")
    print(Task129.retrieve_file("test1.txt"))
    print("---\\n")
    
    # Test case 2: Another valid file
    print("Test 2: Valid JSON file")
    print(Task129.retrieve_file("test2.json"))
    print("---\\n")
    
    # Test case 3: Path traversal attempt
    print("Test 3: Path traversal attempt")
    print(Task129.retrieve_file("../../../etc/passwd"))
    print("---\\n")
    
    # Test case 4: Invalid file extension
    print("Test 4: Invalid file extension")
    print(Task129.retrieve_file("test.exe"))
    print("---\\n")
    
    # Test case 5: Non-existent file
    print("Test 5: Non-existent file")
    print(Task129.retrieve_file("nonexistent.txt"))
    print("---\\n")

if __name__ == "__main__":
    main()

File: dataset/ZeroShot/CLAUDE_claude-sonnet-4-5-20250929/Task129.java

import java.io.*;
import java.nio.file.*;
import java.security.MessageDigest;
import java.util.*;
import java.util.regex.Pattern;

public class Task129 {
    private static final String BASE_DIRECTORY = "./secure_files/";
    private static final Set<String> ALLOWED_EXTENSIONS = new HashSet<>(
        Arrays.asList(".txt", ".json", ".xml", ".csv", ".log")
    );
    private static final long MAX_FILE_SIZE = 10 * 1024 * 1024; // 10MB
    private static final Pattern SAFE_FILENAME_PATTERN = Pattern.compile("^[a-zA-Z0-9_\\\\-\\\\.]+$");
    
    public static String retrieveFile(String fileName) {
        try {
            // Input validation
            if (fileName == null || fileName.trim().isEmpty()) {
                return "Error: File name cannot be empty";
            }
            
            fileName = fileName.trim();
            
            // Check for safe filename pattern
            if (!SAFE_FILENAME_PATTERN.matcher(fileName).matches()) {
                return "Error: Invalid file name format. Only alphanumeric, dash, underscore and dot allowed";
            }
            
            // Check file extension
            String extension = getFileExtension(fileName);
            if (!ALLOWED_EXTENSIONS.contains(extension.toLowerCase())) {
                return "Error: File type not allowed. Allowed types: " + ALLOWED_EXTENSIONS;
            }
            
            // Prevent path traversal attacks
            Path basePath = Paths.get(BASE_DIRECTORY).toAbsolutePath().normalize();
            Path filePath = basePath.resolve(fileName).normalize();
            
            if (!filePath.startsWith(basePath)) {
                return "Error: Path traversal detected. Access denied";
            }
            
            // Check if file exists
            File file = filePath.toFile();
            if (!file.exists()) {
                return "Error: File not found";
            }
            
            // Check if it's a file (not directory)\n            if (!file.isFile()) {\n                return "Error: Invalid file type";\n            }\n            \n            // Check file size\n            if (file.length() > MAX_FILE_SIZE) {\n                return "Error: File size exceeds maximum allowed size";\n            }\n            \n            // Read file content\n            StringBuilder content = new StringBuilder();\n            try (BufferedReader reader = new BufferedReader(new FileReader(file))) {\n                String line;\n                while ((line = reader.readLine()) != null) {\n                    content.append(line).append("\\n");\n                }\n            }\n            \n            return "Success: File retrieved\\nContent:\\n" + content.toString();\n            \n        } catch (IOException e) {\n            return "Error: Unable to read file - " + e.getMessage();\n        } catch (Exception e) {\n            return "Error: Unexpected error - " + e.getMessage();\n        }\n    }\n    \n    private static String getFileExtension(String fileName) {\n        int lastDot = fileName.lastIndexOf('.');
        if (lastDot > 0 && lastDot < fileName.length() - 1) {
            return fileName.substring(lastDot);
        }
        return "";
    }
    
    private static void setupTestEnvironment() {
        try {
            File baseDir = new File(BASE_DIRECTORY);
            if (!baseDir.exists()) {
                baseDir.mkdirs();
            }
            
            // Create test files
            createTestFile("test1.txt", "This is a test file 1");
            createTestFile("test2.json", "{\\"name\\": \\"test\\", \\"value\\": 123}");
            createTestFile("test3.csv", "id,name,age\\n1,John,30\\n2,Jane,25");
        } catch (Exception e) {
            System.err.println("Setup error: " + e.getMessage());
        }
    }
    
    private static void createTestFile(String fileName, String content) throws IOException {
        File file = new File(BASE_DIRECTORY + fileName);
        try (FileWriter writer = new FileWriter(file)) {
            writer.write(content);
        }
    }
    
    public static void main(String[] args) {
        setupTestEnvironment();
        
        System.out.println("=== Secure File Retrieval System ===\\n");
        
        // Test case 1: Valid file retrieval
        System.out.println("Test 1: Valid file retrieval");
        System.out.println(retrieveFile("test1.txt"));
        System.out.println("---\\n");
        
        // Test case 2: Another valid file
        System.out.println("Test 2: Valid JSON file");
        System.out.println(retrieveFile("test2.json"));
        System.out.println("---\\n");
        
        // Test case 3: Path traversal attempt
        System.out.println("Test 3: Path traversal attempt");
        System.out.println(retrieveFile("../../../etc/passwd"));
        System.out.println("---\\n");
        
        // Test case 4: Invalid file extension
        System.out.println("Test 4: Invalid file extension");
        System.out.println(retrieveFile("test.exe"));
        System.out.println("---\\n");
        
        // Test case 5: Non-existent file
        System.out.println("Test 5: Non-existent file");
        System.out.println(retrieveFile("nonexistent.txt"));
        System.out.println("---\\n");
    }
}
